src/libtorch.cc

```diff
@@ -728,6 +732,59 @@ ModelInstanceState::ValidateInputs(const size_t expected_input_cnt)
 
   bool supports_batching = model_state_->MaxBatchSize() > 0;
 
+  NamingConvention naming_convention = NamingConvention::ARGUEMENT;
+
+  // Rules for (non-Dictionary) input tensor names:
+  // 1. Must be in 'allowed_inputs' (arguments in the forward function)
+  // 2. Must follow the naming convention i.e. <name>__<index>
+  // 3. If neither of the above conditions are satisfied, enforce strict
+  // ordering of model inputs.
+  for (size_t i = 0; i < ios.ArraySize(); i++) {
+    triton::common::TritonJson::Value io;
+    RETURN_IF_ERROR(ios.IndexAsObject(i, &io));
+
+    // Validate name
+    std::string io_name;
+    RETURN_IF_ERROR(io.MemberAsString("name", &io_name));
+    auto itr = allowed_inputs.find(io_name);
+    if (itr == allowed_inputs.end()) {
+      naming_convention = NamingConvention::STRICT_ORDERING;
+      break;
+    }
+  }
+
+  // If not, check if inputs follow INDEX
+  if (naming_convention == NamingConvention::STRICT_ORDERING) {
+    for (size_t i = 0; i < ios.ArraySize(); i++) {
+      triton::common::TritonJson::Value io;
+      RETURN_IF_ERROR(ios.IndexAsObject(i, &io));
+
+      // Validate name
+      std::string io_name;
+      RETURN_IF_ERROR(io.MemberAsString("name", &io_name));
+      try {
+        int start_pos = io_name.find(deliminator);
+        if (start_pos == -1) {
+          throw std::invalid_argument("input must follow naming convention");
+        }
+        ip_index = std::atoi(io_name.substr(start_pos + 2).c_str());
+        naming_convention = NamingConvention::INDEX;
+      }
+      catch (std::exception& ex) {
+        LOG_MESSAGE(
+            TRITONSERVER_LOG_WARN,
+            ("input '" + io_name + "' is neither an input argument to the " +
+             "model '" + model_state_->Name() +
+             "' nor does it follow the naming convention i.e. "
+             "<name>__<index>. Falling back to enforcing ordering from model "
+             "configuration.")
+                .c_str());
+        naming_convention = NamingConvention::STRICT_ORDERING;
+        break;
+      }
+    }
+  }
```

Comment:
Put above to a helper function i.e. `GetNamingConvention()` and it should be able to be re-used by output validation with some twists, i.e. passing `allow_inputs` as argument and use empty set for output to skip the check as there is no named output support